### PR TITLE
Allow ContentListener and AsyncContentListener to throw Exception

### LINKED ISSUE #11800: client: Allow AsyncContentListener.onContent to throw checked Exceptions
**Jetty version(s)**
Jetty 12.x

**Enhancement Description**
[`AsyncContentListener` implements `ContentSourceListener.onContentSource`](https://github.com/jetty/jetty.project/blob/a9b2da533f64b923cf06306b5f5faf2c5132f7ce/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java#L181) and delegates to its own hook called `onContent`. Exceptions (or any Throwable really) thrown from onContent are catched and handled inside AsyncContentListener.onContentSource.

Now I have an implementation of `AsyncContentListener.onContent` that writes the content to a file, which can fail with an `IOException`. I would like to let these exceptions bubble up to `onContentSource`, which would handle it nicely. But this is not possible since `IOException` is a checked exception.

There are two workarounds:
One is to wrap the `IOException` in a `RuntimeException` and before rethrowing it.
The other is to completely handle the exception myself, basically copying the `catch` block from `AsyncContentListener .onContentSource`.

### PR DESCRIPTION
**Pull request by lorban**
Closes #11800

### CODE REVIEW

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java [hunk 1]
```diff
@@ -175,7 +169,17 @@ default void onContent(Response response, Content.Chunk chunk, Runnable demander
...
-        void onContent(Response response, Content.Chunk chunk, Runnable demander);
+        /**
+         * Callback method invoked when the response content has been received, parsed and there is demand.
+         * The {@code demander} must be run before this method may be invoked again.
+         *
+         * @param response the response containing the response line data and the headers
+         * @param chunk the chunk received
+         * @param demander the runnable to be run to demand the next chunk
+         * @throws Exception an uncaught exception will abort the response, release the chunk and fail the content source
+         * from which the chunk was returned from
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java]:
```suggestion
         * from which the chunk was read from
```

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java [hunk 2]
```diff
@@ -175,7 +169,17 @@ default void onContent(Response response, Content.Chunk chunk, Runnable demander
      */
     interface AsyncContentListener extends ContentSourceListener
     {
-        void onContent(Response response, Content.Chunk chunk, Runnable demander);
+        /**
+         * Callback method invoked when the response content has been received, parsed and there is demand.
+         * The {@code demander} must be run before this method may be invoked again.
+         *
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:178]: Please add to the javadocs that upon return from the method, the chunk is released.
It is implicit in the doctrine, but does not hurt to repeat it (see javadocs of `ContentListener.onContent()`).

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java [hunk 3]
```diff
@@ -150,21 +150,15 @@ interface ContentListener extends AsyncContentListener
          *
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:151]: Please add that upon returning normally from the method, there is automatic demand.

**Review by sbordet — CHANGES_REQUESTED** [review]
Please also add a test case, if not already present.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java]: Done.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:178]: Done.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java (hunk 3 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:151]: Done.

**Review by sbordet — APPROVED** [review]

**Review by gregw — APPROVED** [review]

**Pull request merged**